### ISSUE TITLE
The question of the accuracy of the algorithm under the iid setting of the mnist dataset

### ISSUE THREAD

**Issue by Alan-JW** [TsingZ0/PFL-Non-IID#12]
Hi， @TsingZ0 
Thank you very much for your contribution to the open-source community. I used “python generate_mnist.py iid - -generate_mnist" to generate the iid dataset under mnist, and then used the statements in "auto_train.sh" to run the corresponding algorithm, but the accuracy of several algorithms seems to have problems, as shown in the following figure:

![7cebd66652af16ec1eca3998f9c8063](https://user-images.githubusercontent.com/74579183/174423045-2d995569-79c3-46d6-8c39-0a64401f3664.png)
![image](https://user-images.githubusercontent.com/74579183/174423102-29601f7e-8295-4291-8fb7-56c0ac1db441.png)
![image](https://user-images.githubusercontent.com/74579183/174423111-681c8770-5c9a-4296-a89e-e9e04666d2ee.png)
![image](https://user-images.githubusercontent.com/74579183/174423129-92d3baf8-cbc0-4fec-bb44-b484783e86c8.png)

Thank you very much for your answer！

**Comment by TsingZ0** [comment]
Hi, @Alan-JW .

For **pFedMe**, the hyper-parameter setting matters. The hyper-parameters in "auto_train.sh" is just examples, you need to tune them by yourself. 

Using 
`python -u main.py -lr 0.1 -lbs 16 -nc 20 -jr 1 -nb 10 -data mnist -m cnn -algo pFedMe -gr 2000 -did 6 -lrp 0.01 -bt 1 -lam 15`
or
`python -u main.py -lr 0.005 -lbs 16 -nc 20 -jr 1 -nb 10 -data mnist -m cnn -algo pFedMe -gr 2000 -did 6 -lrp 0.01 -bt 1 -lam 15` 
works well. 

**Hyper-parameter tuning** is also required for other algorithms (such as the mentioned FedAMP and FedMTL).  

Besides, the accuracy of most personalized algorithms is actually worse than FedAvg in the IID setting. Since they are designed for the Non-IID setting and each client model can only benefit from a few other client models, which is not suitable in the IID setting. 

*You can further study this phenomenon by varying the `alpha` in `dataset_utils.py` for Dirichlet distribution to control the degree of Non-IID in the **practical Non-IID** setting.*

In the IID setting, for **FedFomo**, setting `self.val_ratio = 0.2` may reduce the number of training data samples, which makes the local model hard to train well because of the lack of data for each label. You can decrease `self.val_ratio` or use the trainloader as the val_loader (with `self.val_ratio = 0`).

**Comment by TsingZ0** [comment]
Hi, @Alan-JW 

I checked the code again, and address this issue by setting shuffle=True in ./dataset/utils/dataset_utils.py.
Besides, I also update other codes in ./dataset/. Please check it.

Thank you for your help!

Please refer to #14 for the results on MNIST in the IID setting :-).

**Comment by Alan-JW** [comment]
Thank you very much for your answer！
